### PR TITLE
command: add display-width/display-height property

### PR DESCRIPTION
**Pull request by Dudemanguy**
For some reason, this never existed before. Add VOCTRL_GET_DISPLAY_RES
and use it to obtain the current display's resolution from each
vo/windowing backend if applicable. Users can then access the current
display resolution as display-width and display-height as per the client
api. Note that macOS/cocoa was not attempted in this commit since the
author has no clue how to write swift.

Sidenote: I didn't test windows or rpi.

~Edit: Now this PR covers more video/display properties.~ Not anymore.

### CODE REVIEW

**Comment by haasn** [comment]
Why does this need to go via the display backend? Doesn't this already exist in the form of `vo->dwidth/dheight`? Can't you just return that directly from vo.c?

Also, it might be important to distinguish this from the *scaled* output size. For example, if I am watching a 640x480 video on a 1920x1080 monitor, your `display-width/display-height` will return 1920x1080, but the video will be displayed as 1440x1080.

I think it would be useful to add both of those, to avoid confusion. And neither of them should require going via the vo backend.

**Comment by Dudemanguy** [comment]
This is meant to return the display resolution. `vo->dwidth/dheight` returns the video size. The scaled size case you mentioned should already be covered by `vo->dwidth/dheight`.

**Comment by haasn** [comment]
No, on my end `vo->dwidth/dheight` returns the size of the window, *not* the size of the scaled video.

**Comment by Dudemanguy** [comment]
>No, on my end vo->dwidth/dheight returns the size of the window, not the size of the scaled video.

Ah yes, you're correct. Important distinction there.

>Though the question is whether or not you care about the window size or the monitor size, when you refer to 'display'. I assume you mean the window size.

I think I'm just following the existing convention here. Nearly every current mpv property titled `display-*` refers to actual properties of the monitor. The documentation mostly refers to "display" as in "monitor".

**Comment by haasn** [comment]
Okay. Well, I suppose that makes more sense then. (Though I somewhat question what this is useful for)

Maybe we should take this opportunity to just add all the properties one could possibly want, and in doing so, make sure we come up with a non-confusing name for each. Specifically, we need the following:

> Native file resolution
> Filter chain output resolution
> Scaled video output resolution
> Window size
> Display size

I think 'window' is a natural fit for the window size, but what do we call the "Scaled video output resolution" field? My idea would be "target-width/target-height". I don't think it currently has any analogs.

As an aside, we probably also want access to "target-rect" and "source-rect" or something, which returns the equivalent of `src_rect` and `dst_rect` in `vo_get_src_dst_rects`.

There's also the question of whether or not "Scaled video output resolution" gives the *cropped* values or the *uncropped* values. I think it would probably make the most sense to have it match the semantics of the "Filter chain output resolution" (i.e. what is currently `video-out-params/{w,h}`), which I think is the uncropped size.

This, plus access to the src/dst rects (which give you not just the cropped size but also the crop coordinates) would give full coverage of all the relevant information.

**Comment by haasn** [comment]
On further thought, the target resolution would be redundant with having a single "scale factor", which I think already exists (window-scale?). I guess to summarize, here's my concrete proposal:

Property | Semantics
---------|----------
`video-params/{w,h}` | Native file resolution, as encoded
`video-params/{dw,dh}` | Native file resolution, with aspect ratio correction
`video-params/crop` | See `video-out-params/crop`. Included only for consistency.
`video-out-params/{w,h}` | Output of the filter chain, as encoded
`video-out-params/{dw,dh}` | Output of the filter chain, with aspect ratio correction
`video-out-params/crop` | Portion of the (aspect ratio corrected) image that is being displayed, guaranteed to be a sub-crop of (dw, dh)
*`target-scale` (maybe)* | *The effective end-to-end scaling ratio between `video-out-params` and the image as displayed on-screen. I think a single number suffices here because I don't believe mpv allows monitor PAR to differ from the image PAR? This may already be redundant with `target-scale`, I haven't checked.*
`window-{width,height}` | Size of the window being rendered into
`window-crop` | The rect of the window that contains the scaled image. This is *almost* redundant with `video-out-params/crop` + scale factor, but not quite, because the output can be panned inside the window.
`display-{width,height}` | Resolution of the monitor that mpv is displayed on on

This set of properties allows you to derive anything else you'd be interested in:
* Scaled video output resolution = `video-out-params/{dw,dh}` × scale factor
* Scaled video output crop = `video-out-params/crop` × scale factor
* Transform from window pixel coordinate to native image coordinate (for e.g. inserting filters): Divide pixel space by `window-crop` and multiply by `video-params/crop`.

**Comment by Dudemanguy** [comment]
Okay so as per the above discussion, I went ahead and expanded this PR to cover various video/display properties that might be nice to have.

TODO:
- [ ] `display-width` / `display-height` (I know it seems silly but it'd be nice to have for at least one of my scripts)
- [ ] `window-width` / `window-height`
- [ ] `video-scaled-width` / `video-scaled-height`
- [ ] `video-out-params/crop` (and other crop stuff)

`target-scale` is covered by `current-window-scale` so no need to implement that. There are edge cases where `video-out-params/{dw,dh}` × `current-window-scale` won't return the correct values so I did create a specific `video-scaled-width/height` properties for that. For example consider overriding the video aspect on some video to something else and then resize it manually. `video-out-params/{dw,dh}` × `current-window-scale` wouldn't be right in that case.

**Comment by haasn** [comment]
Another edge case I didn't consider above is `--no-keepaspect`, in which case `video-scaled-width/height` ends up effectively equal to `window-width/height`, but the fact that this is extra logic justifies adding the property.

**Comment by Dudemanguy** [comment]
Just added `window-crop/{x0,x1,y0,y1}` and `{video-out-params,video-params}/{x0,x1,y0,y1}`. It gets kind of confusing to explain but `window-crop` is the actual location of the scaled video image with respect to the **mpv window**. So all of the values returned here will always be contained inside whatever the actual window size. On the other hand the `video-out-params` stuff returns its values with respect to the **video size**. So these values with always be contained within `dwidth` and `dheight`. I think that covers all of that video/window stuff unless someone else can think of something.

Implementation note: I took the easy way out and added `x0`, `x1`, `y0`, and `y1` as additional sub properties of `video-out-params` and `video-params` instead of doing `video-out-params/crop` (or something similar). Having sub properties of a sub property isn't really a thing (`playlist` and other stuff actually map an array to a node for their entries which isn't applicable here). Maybe it could be cleverly implemented in the m_property stuff, but I didn't try it.

**Comment by avih** [comment]
Does this PR take into account multi-monitor setups? It can be painful to handle at the win32 code.

Also, on Windows, `get_working_area` is the working area of one display - where the majority of the window is currently, after removing fixed taskbar and such. It's definitely not the monitor size in any way.

File: player/command.c [hunk 1]
```diff
@@ -2404,6 +2494,23 @@ static int mp_property_vsync_jitter(void *ctx, struct m_property *prop,
...
+        return M_PROPERTY_UNAVAILABLE;
+    struct mp_rect res;
+    if (vo_control(vo, VOCTRL_GET_DISPLAY_RES, &res) <= 0)
+        return M_PROPERTY_UNAVAILABLE;
+    if (strcmp(prop->name, "display-width") == 0) {
+        return m_property_int_ro(action, arg, mp_rect_w(res));
+    } else {
+        return m_property_int_ro(action, arg, mp_rect_h(res));
+    }
+}
```
Comment (avih) [inline player/command.c:2512]: If you're going to unconditionally turn a 4-values rectangle into a two-values size, then the API has a redundancy. It should instead pass only two values: width and height.

**Comment by avih** [comment]
How many of these new properties duplicate values at the `osd-dimensions`?

How many of these new properties can be trivially calculated from other existing properties?

But I think even before we review the actual PR, I still don't see the value in it. It's not used by mpv itself, and I don't see an obvious use case where a script or client would have good use for these.

So why do we add these properties, idiosyncracies between various VOs and maintenance burden?

E.g. `display-width` on windows is the working area on the current monitor (subject to change if the user manually moves the window to another monitor - without mpv being notified about it), after reducing the taskbar, and probably after taking desktop scaling into account.

In other words, some of these values simply speak in different terms between VOs.

**Comment by Dudemanguy** [comment]
>How many of these new properties can be trivially calculated from other existing properties?

Not many? I mean there's a whole discussion on this above. That said. `osd-dimensions` acts completely different than I thought it did so I'll double check how that actually works later. It doesn't just return `vo->dwidth/dwheight` (rather something else totally) from the VO so I wouldn't expect it to always be the same, but I still need to double check the logic there.

**Comment by haasn** [comment]
> At least the docs say it's not supposed to always be equal to the window size but I haven't found an example where that's not the case yet.

This may have been future-proofing against CSDs?

**Comment by Dudemanguy** [comment]
So I hunted it down and it turns out that said OSD values originate from a [mp_get_src_dst_rects] call (https://github.com/mpv-player/mpv/blob/6265724f3331e3dee8d9ec2b6639def5004a5fa2/video/out/vo.c#L1243) in vo.c which directly pass `vo->dwidth` and `vo->dheight`. In said [function](https://github.com/mpv-player/mpv/blob/6265724f3331e3dee8d9ec2b6639def5004a5fa2/video/out/aspect.c#L160), those values directly correspond to the osd `w` and `h` values. So in fact yes, it will always be exactly the same despite the docs saying otherwise. How confusing. I'll remove that commit from here.

**Comment by avih** [comment]
```
``video-scaled-width``, ``video-scaled-height``
    Scaled video size. Note that is not the same as the size of the mpv window.
    It only measures the size of the actual video after it is scaled and resized.
```

I Find the description unclear. What do "only" and "actual" meant to clarify? Is "resized" supposed to take into account video-zoom and panscan? (it doesn't seem to)

Also, I find the `window-crop` values and the new properties at `video-params` highly obscure, and I still can't think of a use case for them.

Especially when `osd-dimensions` already gives the full info in display units - the window size, and the final video size and position, in display units relative to the window (via the margins values - which can be negative). Which means that all the crop and intersections/positioning between the window and the video (final scaled value/position in display units) can be deduced trivially from it.

I also still think the display-resolution property is highly problematic, ambiguously documented, and pretty much useless when decorations are involved.

Nevertheless, other than not being particularly useful IMHO and potential maintenance burden, I also don't see it doing any real harm. So I lift my objection and let others decide.

**Comment by Dudemanguy** [comment]
On second thought, actually everything (except for `display-width/height`) is useless because osd-dimensions already tells you everything. I really the docs didn't lie to me about the "which can be different from the window size in some cases" part... (I'll get rid of that later).

>I also still think the display-resolution property is highly problematic, ambiguously documented, and pretty much useless when decorations are involved.

What on earth does decorations have to do with a monitor size?

>Also, I think a user will have to poll it, because I don't think currently it can be observed in a reliable way because mpv is unaware when the window moves to a different monitor (or moved at all).

No they don't. It's exactly like every other display-* property. It's already the responsibility of the windowing backend to signal these things to the VO. We do exactly the same thing for `display-fps` for example. That said, it is, of course, the most intrusive change since it has to go through a VOCTRL.

**Comment by avih** [comment]
> What on earth does decorations have to do with a monitor size?

The only value which the display dimensions have is in helping you make a decision about the window size when you want to resize it or realize if it fits on screen, but if you don't know how much space the decorations take additionally, then you still can't decide, i.e. knowing the dimensions didn't help you make a better decision.

> It's already the responsibility of the windowing backend to signal these things to the VO

Sure, except that as far as I know currently none of the VOs actually do that with the display dimensions. The code which you added in this PR only supports polling this value from the VO. And adding active push on monitor change can be non trivial, and more maintenance burden for a feature which no one ever requested.

**Comment by Dudemanguy** [comment]
>The only value which the display dimensions have is in helping you make a decision about the window size when you want to resize it or realize if it fits on screen, but if you don't know how much space the decorations take additionally, then you still can't decide, i.e. knowing the dimensions didn't help you make a better decision.

This seems more like your personal opinion. What if someone just doesn't care about decorations and wants to resize/whatever based on the display resolution properties?

>Sure, except that as far as I know currently none of the VOs actually do that with the display dimensions.

Wayland keeps track of the monitor the mpv window is currently on, so yes it does. I'd assume mac and windows do as well since, as far as I know, both OSes are also capable of handling multiple outputs properly. If the window changes monitors, the backend is supposed to send a `VO_EVENT_WIN_STATE`. It's already required to keep track of geometry and signal the VO when something changes.

**Comment by avih** [comment]
> This seems more like your personal opinion

Very true. In fact, I already started to edit and mention explicitly that personally I can't see a value other than this (like I said explicitly the previous times I mentioned value), but you posted before I completed the edit.

Nevertheless, you're adding these features, so I'd think you'd have at least some use cases in mind, but so far you didn't give any real reason to adding these features, or sample use cases which could benefit from having them.

**Comment by avih** [comment]
For reference, this lua script will print the window size and and video size and position relative to the top-left corner of the window - after taking into account stretching, zoom levels, pan-and-scan, etc, where some or all of the video boundaries could be outside the window.

```lua
local msg = require("mp.msg")
local utils = require("mp.utils")

mp.observe_property("osd-dimensions", "native", function(n, od)
    local window = utils.format_json({
        w = od.w,
        h = od.h
    })
    -- some/all of these can be outside the window 0,0,w,h
    local video_pos = utils.format_json({
        x = od.ml,
        y = od.mt,
    })
    local video = utils.format_json({
        w = od.w - od.mr - od.ml,
        h = od.h - od.mb - od.mt
    })

    msg.info("Window:", window, '@{"x":0,"y":0)')
    msg.info(" Video:", video, "@" .. video_pos)
    msg.info()
end)

```

This allows to trivially calculate all the margins, croppings, intersections, visible parts, etc between the video and the window.

**Comment by Dudemanguy** [comment]
So it's officially back to just display-width/height again.

File: player/command.c (hunk 1 above)
Comment (Dudemanguy) [reply player/command.c:2422]: mp_rect was chosen mainly since it's convenient. I guess I could return a int[] or something here instead.

Comment (Dudemanguy) [reply player/command.c]: Should be okay now. Just mac needs updating later.

File: DOCS/man/input.rst [hunk 2]
```diff
@@ -2529,6 +2529,11 @@ Property list
 ``vsync-jitter``
     Estimated deviation factor of the vsync duration.
 
+``display-width``, ``display-height``
+    The current display's horizontal and vertical resolution in pixels. Whether
```
Comment (avih) [inline DOCS/man/input.rst:2533]: Are you 100% sure it's always pixels in all VO backends? Documentation for `osd-dimensions` say
```
usually pixels, but may be scaled pixels with VOs like xv
```
Did you try to confirm that it's still pixels in xv where this scaling applies? Also, windows and OSX also have display scaling, and I don't know if the expected value is the virtual resolution or the actual pixels on screen.

File: video/out/w32_common.c [hunk 3]
```diff
@@ -1768,6 +1768,11 @@ static int gui_thread_control(struct vo_w32_state *w32, int request, void *arg)
         update_display_info(w32);
         *(double*) arg = w32->display_fps;
         return VO_TRUE;
+    case VOCTRL_GET_DISPLAY_RES: ;
+        RECT r = get_screen_area(w32);
```
Comment (avih) [inline video/out/w32_common.c:1772]: Nice. I was about to comment that `get_screen_area` is likely more aligned with the docs than `get_working_area`, but you changed it on your own :)

File: DOCS/man/input.rst (hunk 2 above)
Comment (Akemi) [reply DOCS/man/input.rst:2533]: the actual. since we don't work with pt anywhere.

Comment (avih) [reply DOCS/man/input.rst:2533]: OK, I also tested on windows, and both `get_working_area` and `get_screen_area` return a size in pixels, apparently regardless of Windows Scaling settings (Windows 10: right-click desktop -> Display Settings -> Advanced scaling settings: tried both custom scale value of 125%, and also tried turning off "Fix scaling for apps" - result remains in pixels).

Comment (Dudemanguy) [reply DOCS/man/input.rst:2533]: Yeah, it's the actual pixels. The same is true for xv/wlshm and similar VOs.
